native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js

```diff
@@ -34,92 +33,114 @@ const {
+        this.setState({ inTree: true, expandAnimationInProgress: true });
+      }
     }
   }
 
-  setExpandedHeight(event, expandedImmediately) {
-    const { height } = event.nativeEvent.layout;
-    this.setState({
-      // if the component has started expanded, we should now set the height of the container
-      // why on earth does a value of 1 instead of 0 make this break!
-      height: new Animated.Value(expandedImmediately ? height : 0),
-      heightSet: true,
-      expandedHeight: height,
-    });
+  onHeightAnimationComplete() {
+    if (typeof this.props.onAnimationComplete === 'function') {
+      this.props.onAnimationComplete();
+    }
+    if (this.state.height._value - 0.01 < 0.00001) { // eslint-disable-line no-underscore-dangle
+      this.setState({ inTree: false });
+    }
   }
 
-  resize() {
-    // overflow: hidden is not properly supported on Android,
-    // which causes animation to look shoddy.
-    // See https://facebook.github.io/react-native/releases/0.49/docs/layout-props.html#overflow
-    const animationDuration = Platform.OS === 'ios' ? this.props.animationDuration : 0;
-
-    const finalValue = this.state.expanded ? this.state.expandedHeight : this.state.collapsedHeight;
-
-    Animated.timing(
-      this.state.height,
-      {
-        toValue: finalValue,
-        duration: animationDuration,
-      },
-    ).start();
+  setExpandedHeight(event) {
+    const { height } = event.nativeEvent.layout;
+    if (height !== 0) {
+      if (!this.state.expandAnimationInProgress) {
+        this.setState({
+          height: this.state.expanded ? new Animated.Value(height) : new Animated.Value(collapsedHeight),
+        });
+      } else {
+        this.setState({ expanded: true, expandAnimationInProgress: false });
+        Animated.timing(this.state.height, {
+          toValue: height,
+          duration: this.animationDuration,
+          delay: this.props.expandDelay,
+        }).start(this.onHeightAnimationComplete);
+      }
+    }
   }
 
   render() {
-    const { expanded, style, ...rest } = this.props;
+    const {
+      animationDuration, expanded, children, expandDelay, collapseDelay, onAnimationComplete, ...rest
+    } = this.props;
 
     const { height } = this.state;
 
-    return (
-      <Animated.View
-        style={{
-          ...style,
+    return this.state.inTree ?
+      (
+        <View
+          {...rest}
+        >
+          <Animated.View
+            style={{
           overflow: 'hidden',
           height,
         }}
-        {...rest}
-      >
-        <View
-          onLayout={event => (this.state.heightSet ? undefined : this.setExpandedHeight(event, expanded))}
-        >
-          {this.props.children}
+          >
+            <View
+              onLayout={event => (this.setExpandedHeight(event))}
+            >
+              {this.props.children}
+            </View>
+          </Animated.View>
         </View>
-      </Animated.View>
-    );
+      ) : null;
```

Comment:
So per the CHANGELOG this should be removing itself from the tree, but this actually removes the children not it self

Reply:
Not sure what you mean here. The render function literally returns null.
The View-count shown below always reverts to 11 after the animation has completed. It also shows 11 views if I delete the `AnimateHeight` component from the story completely.

![null](https://user-images.githubusercontent.com/30267516/32835117-d4eac8b8-c9fc-11e7-9456-2ec89f0d5ba0.gif)

Reply:
The component still renders right? It just renders no children

Reply:
So the component shown above is a `View` containing a `BpkAnimateHeight` and a `BpkButton`. Once the `BpkAnimateHeight` has collapsed it becomes null, so all you're seeing is a `Viiew` containing the `BpkButton`